monkey/infection_monkey/island_api_client/http_client.py

```diff
@@ -0,0 +1,134 @@
+import functools
+import logging
+from enum import Enum, auto
+from typing import Any, Dict, Optional
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3 import Retry
+
+from common.common_consts.timeouts import MEDIUM_REQUEST_TIMEOUT
+from common.types import JSONSerializable
+
+from .island_api_client_errors import (
+    IslandAPIConnectionError,
+    IslandAPIError,
+    IslandAPIRequestError,
+    IslandAPIRequestFailedError,
+    IslandAPITimeoutError,
+)
+
+logger = logging.getLogger(__name__)
+
+# Retries improve reliability and slightly mitigate performance issues
+RETRIES = 5
+
+
+class RequestMethod(Enum):
+    GET = auto()
+    POST = auto()
+    PUT = auto()
+
+
+def handle_island_errors(fn):
+    @functools.wraps(fn)
+    def decorated(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except IslandAPIError as err:
+            raise err
+        except (requests.exceptions.ConnectionError, requests.exceptions.TooManyRedirects) as err:
+            raise IslandAPIConnectionError(err)
+        except requests.exceptions.HTTPError as err:
+            if 400 <= err.response.status_code < 500:
+                raise IslandAPIRequestError(err)
+            elif 500 <= err.response.status_code < 600:
+                raise IslandAPIRequestFailedError(err)
+            else:
+                raise IslandAPIError(err)
+        except TimeoutError as err:
+            raise IslandAPITimeoutError(err)
+        except Exception as err:
+            raise IslandAPIError(err)
+
+    return decorated
+
+
+class HTTPClient:
+    def __init__(self, retries=RETRIES):
+        self._session = requests.Session()
+        retry_config = Retry(retries)
+        self._session.mount("https://", HTTPAdapter(max_retries=retry_config))
+        self._api_url: Optional[str] = None
+
+    def connect(self, island_server: str):
```

Comment:
Island server is `SocketAddress` not `str`.
```suggestion
    def connect(self, island_server: SocketAddress):
```

Reply:
Why didn't mypy complain though 🤔 ?

Reply:
Maybe because it can be cast to string and is used as such?

Reply:
Not sure. We need to check that.